File: eprofiler-core/src/main/java/cz/encircled/eprofiler/Profiler.java
package cz.encircled.eprofiler;

import cz.encircled.eprofiler.registry.MethodDescriptor;
import cz.encircled.eprofiler.registry.MethodRegistry;

import java.lang.management.ManagementFactory;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;

/**
 * @author Kisel on 24.05.2016.
 */
public class Profiler {

//    private static ArrayBlockingQueue<MethodState> states = new ArrayBlockingQueue<>(16);

    static ThreadLocal<MethodState> state = new ThreadLocal<MethodState>() {
        @Override
        protected MethodState initialValue() {
            return null;
        }
    };
    private static AtomicLong idCounter = new AtomicLong(1L);

    public static MethodState methodStart(long id) {
        long cpuAtStart = ManagementFactory.getThreadMXBean().getCurrentThreadCpuTime();

        MethodDescriptor descriptor = MethodRegistry.get(id);

        MethodState state = Profiler.state.get();
        if (state == null) {
            state = new MethodState();
        } else {
            MethodState sibling = findRepeatedSibling(state.children, id);
            if (sibling != null) {
                // Simple loop
                sibling.repeats++;
                Profiler.state.set(sibling);
                sibling.starts.add(Timer.now);
                sibling.cpuBefore = cpuAtStart;
                return sibling;
            } else {
                MethodState nested = new MethodState();
                state.addNested(nested);
                nested.parent = state;
                state = nested;
            }
        }
        Profiler.state.set(state);
        state.executionId = idCounter.getAndIncrement();
        state.starts.add(Timer.now);
        state.descriptor = descriptor;
        state.cpuBefore = cpuAtStart;
        return state;
    }

    private static MethodState findRepeatedSibling(List<MethodState> states, long id) {
        for (MethodState state : states) {
            if (state.descriptor.id == id) {
                return state;
            }
        }
        return null;
    }

}

File: eprofiler-core/src/main/java/cz/encircled/eprofiler/Profiler.kt
package cz.encircled.eprofiler

import java.util.concurrent.atomic.AtomicLong

/**
 * @author Kisel on 24.05.2016.
 */
object Profiler {

    //    private static ArrayBlockingQueue<MethodState> states = new ArrayBlockingQueue<>(16);

    var state: ThreadLocal<MethodState> = object : ThreadLocal<MethodState>() {
        override fun initialValue(): MethodState? {
            return null
        }
    }
    private val idCounter = AtomicLong(1L)

    fun methodStart(): MethodState {
        var state: MethodState? = Profiler.state.get()
        if (state == null) {
            state = DefaultMethodState(0)
            state.id = idCounter.andIncrement
            state.setStartTime(Timer.now)
            Profiler.state.set(state)
            return state
        } else {
            val nested = DefaultMethodState(0)
            nested.id = idCounter.andIncrement
            nested.setStartTime(Timer.now)
            nested.parent = state
            state.addNested(nested)
            Profiler.state.set(nested)
            return nested
        }
    }

}
